File: packages/kits/NeuronKit/Sources/NeuronKit/Dreaming/EstateDreamingReader.swift
import Foundation
import GeniusLocusKit

/// Production adapter that binds `DreamingSubstrateReader` to a live
/// GeniusLocusKit estate (NEURONKIT_SPEC § 3.1).
///
/// `DreamingSubstrateReader` is the read seam the daemon uses during a
/// cycle. This adapter satisfies it by delegating to the three substrate
/// reads GLK exposes through `recentRecallTraces(in:since:now:)`,
/// `drainDreamingItems(for:)`, and `allTunnels(in:)` — all B-1-compliant
/// calls through the public GeniusLocusKit verb surface.
///
/// ── Why this lives in NeuronKit, not GeniusLocusKit ──────────────────
/// `DreamingSubstrateReader` is declared here in NeuronKit. A conforming
/// type must import NeuronKit. GeniusLocusKit is a dependency of NeuronKit
/// (GLK sits below NK in the stack), so GLK cannot import NK without
/// creating a circular package dependency. NeuronKit is the only package
/// that can see both the protocol and the GLK estate surface, making it
/// the natural home for this adapter.
///
/// ── Co-occurrence algorithm (v2, drain-fed) ──────────────────────────
/// v2 derives co-recall pairs from the estate's dreaming queue rather
/// than the drawer graph. Each drained window is the drawer-ID set from
/// one recall event that co-recalled ≥ 2 drawers (a `DreamingItem`
/// payload written by the recall verb when it enqueues the dreaming job).
/// The daemon enumerates unordered pairs within each window and bumps
/// `coRecallCounts` once per pair per window (not per cycle) so counts
/// accumulate across drain events. `drainDreamingWindow()` delegates to
/// `GeniusLocusKit.drainDreamingItems(for:)`, which drains the queue and
/// replies Done to consumed jobs before returning.
public struct EstateDreamingReader: DreamingSubstrateReader {

    private let handle: EstateHandle
    private let kit: GeniusLocusKit

    /// Construct an adapter over the addressed estate.
    ///
    /// - Parameters:
    ///   - handle: the estate to read from.
    ///   - kit: the GeniusLocusKit actor that owns the estate registry.
    public init(handle: EstateHandle, kit: GeniusLocusKit) {
        self.handle = handle
        self.kit = kit
    }

    // MARK: - DreamingSubstrateReader

    /// Recall-trace rows in the `[since, now]` reward window.
    /// Delegates to `GeniusLocusKit.recentRecallTraces(in:since:now:)`.
    public func recentRecallTraces(since: Date, now: Date) async throws -> [RecallTraceItem] {
        try await kit.recentRecallTraces(in: handle, since: since, now: now)
    }

    /// Drained dreaming-queue windows for the estate.
    ///
    /// Each inner array is the set of drawer IDs from one `DreamingItem`
    /// (one recall event that co-recalled ≥ 2 drawers). Returns an empty
    /// array when the dreaming queue has not been mounted for this estate
    /// yet (no recall has fired since open). Delegates to
    /// `GeniusLocusKit.drainDreamingItems(for:)`.
    public func drainDreamingWindow() async throws -> [[String]] {
        try await kit.drainDreamingItems(for: handle)
    }

    /// Existing tunnels for duplicate suppression.
    /// Delegates to `GeniusLocusKit.allTunnels(in:)`.
    public func existingTunnels() async throws -> [Tunnel] {
        try await kit.allTunnels(in: handle)
    }

    /// All non-retired dreamed tunnels (T13 / ADR-021 Phase 7).
    ///
    /// Fetches the active-tunnel set from GLK and filters to those with
    /// `isDreamed == true`. Declared tunnels (`isDreamed == false`) are never
    /// returned so OMEGA can never retire them (§ 12.8 guard). Delegates to
    /// `GeniusLocusKit.allActiveTunnels(in:)`.
    public func dreamedActiveTunnels() async throws -> [Tunnel] {
        let active = try await kit.allActiveTunnels(in: handle)
        return active.filter { $0.isDreamed }
    }
}

File: packages/kits/NeuronKit/rust/src/estate_dreaming_reader.rs
// estate_dreaming_reader.rs — Rust parity of
// `NeuronKit/Sources/NeuronKit/Dreaming/EstateDreamingReader.swift`.
//
// Production adapter that implements `DreamingSubstrateReader` over a
// GeniusLocusKit estate handle. The adapter calls the GLK coordinator's
// `drain_dreaming_items` to drain pending dreaming-queue windows (T8 v2);
// the v1 snapshot-based approach (co-occurrence observations from drawer
// pairs) is deleted — `DrainDreamingWindow` is the only candidate source.
//
// Architecture note: same layering rationale as Swift. `EstateDreamingReader`
// lives in NeuronKit because it needs to implement `DreamingSubstrateReader`
// (declared in `dreaming_cycle.rs`) AND call genius_locus_kit coordinator
// methods. NeuronKit already depends on both; genius_locus_kit does not
// depend on neuron_kit, so there is no circular dependency.
//
// B-1 compliance: all estate reads route through genius_locus_kit's
// EstateCoordinator surface — no direct locus_kit storage calls.

use genius_locus_kit::{EstateCoordinator, EstateHandle, VerbDispatchError};

use crate::dreaming_cycle::{
    DreamingSubstrateReader, DreamingTunnelItem, RecallTraceItem as CycleRecallTraceItem,
    TunnelLink,
};

/// Snapshot-based production adapter for `DreamingSubstrateReader`.
///
/// `recent_recall_traces` and `existing_tunnels` are snapshotted from the
/// estate at construction via `EstateDreamingReader::new`. `drain_dreaming_window`
/// is NOT snapshotted — it calls the GLK coordinator's `drain_dreaming_items`
/// lazily when the dreaming cycle calls the trait method (T8 v2 drain-fed approach).
///
/// `recent_recall_traces` and `existing_tunnels` remain snapshot-based because
/// they are pure reads (no queue mutation); `drain_dreaming_window` is stateful
/// (it pops jobs from the queue) so it must be a live call, not a pre-fetch.
///
/// `now_epoch_secs` captures the construction instant (epoch-seconds). The
/// dreaming cycle injects the clock once at reader construction so the drain
/// telemetry is deterministic — no SystemTime reads inside the reader.
pub struct EstateDreamingReader<'a> {
    traces: Vec<CycleRecallTraceItem>,
    tunnels: Vec<TunnelLink>,
    /// Dreamed-active tunnels for OMEGA retire evaluation (T13 / ADR-021 Phase 7).
    /// Snapshotted at construction via `coordinator.all_active_tunnels`, then
    /// filtered to `is_dreamed() == true`. Pre-fetching keeps `dreamed_active_tunnels()`
    /// a cheap `Vec::clone()` inside the cycle — consistent with how `traces` and
    /// `tunnels` are handled.
    dreamed_active: Vec<DreamingTunnelItem>,
    /// Live coordinator reference for the drain call.
    coordinator: &'a EstateCoordinator,
    /// The estate whose dreaming queue this reader drains.
    handle: EstateHandle,
    /// Injected construction timestamp used for drain telemetry.
    now_epoch_secs: f64,
}

impl<'a> EstateDreamingReader<'a> {
    /// Construct the adapter by snapshotting recall traces and tunnels from
    /// the addressed estate through the GeniusLocusKit coordinator surface.
    ///
    /// `since` and `now` are ISO8601 strings bounding the recall-trace
    /// reward window (both inclusive). `now_epoch_secs` is the same instant
    /// in epoch-seconds — used for drain telemetry. Dreaming-queue windows are
    /// NOT snapshotted here — they are drained lazily via `drain_dreaming_window`.
    ///
    /// All reads go through `coordinator.recent_recall_traces` and
    /// `coordinator.all_tunnels` — B-1 compliant.
    pub fn new(
        coordinator: &'a EstateCoordinator,
        handle: &EstateHandle,
        since: &str,
        now: &str,
        now_epoch_secs: f64,
    ) -> Result<Self, VerbDispatchError> {
        let raw_traces = coordinator.recent_recall_traces(handle, since, now)?;
        let traces = raw_traces
            .into_iter()
            .map(|t| CycleRecallTraceItem {
                target: t.target.clone(),
                used: t.used(),
            })
            .collect();

        let raw_tunnels = coordinator.all_tunnels(handle)?;
        let tunnels = raw_tunnels
            .into_iter()
            .map(|t| TunnelLink {
                source_drawer_id: t.source_drawer_id.clone(),
                target_drawer_id: t.target_drawer_id.clone(),
            })
            .collect();

        // Snapshot dreamed-active tunnels for OMEGA retire evaluation (T13 / ADR-021 Phase 7).
        // `all_active_tunnels` excludes retired tunnels (bit 13 clear); filtering to
        // `is_dreamed() == true` enforces the §12.8 guard (declared tunnels never retired).
        // `source_drawer_id` and `target_drawer_id` are Options on the substrate type
        // (room-level tunnels have None endpoints); drawer-pair tunnels always have Some.
        // We skip room-level tunnels (None endpoints) — they cannot be dreamed and are
        // never in the retire population.
        let raw_active = coordinator.all_active_tunnels(handle)?;
        let dreamed_active = raw_active
            .into_iter()
            .filter(|t| t.is_dreamed())
            .filter_map(|t| {
                // Only drawer-pair tunnels (both endpoints Some) enter OMEGA.
                // Room-level tunnels (either endpoint None) are skipped.
                match (&t.source_drawer_id, &t.target_drawer_id) {
                    (Some(src), Some(tgt)) => Some(DreamingTunnelItem {
                        id: t.id.clone(),
                        source_drawer_id: src.clone(),
                        target_drawer_id: tgt.clone(),
                    }),
                    _ => None,
                }
            })
            .collect();

        Ok(EstateDreamingReader {
            traces,
            tunnels,
            dreamed_active,
            coordinator,
            handle: handle.clone(),
            now_epoch_secs,
        })
    }
}

impl<'a> DreamingSubstrateReader for EstateDreamingReader<'a> {
    fn recent_recall_traces(&self) -> Vec<CycleRecallTraceItem> {
        self.traces.clone()
    }

    /// Drain pending dreaming-queue windows from the estate (T8 v2).
    ///
    /// Each inner Vec is the set of drawer IDs from one DreamingItem — a
    /// single recall event that co-recalled ≥ 2 distinct drawers. Returns an
    /// empty Vec when the queue has no pending jobs.
    ///
    /// Delegates to `coordinator.drain_dreaming_items`, which drains the queue
    /// and replies Done to the consumed jobs — drain-once semantics. Matches
    /// Swift `EstateDreamingReader.drainDreamingWindow()`.
    fn drain_dreaming_window(&self) -> Vec<Vec<String>> {
        self.coordinator
            .drain_dreaming_items(&self.handle, self.now_epoch_secs)
            .unwrap_or_default()
    }

    fn existing_tunnels(&self) -> Vec<TunnelLink> {
        self.tunnels.clone()
    }

    /// All non-retired dreamed tunnels for OMEGA retire evaluation (T13 / ADR-021 Phase 7).
    ///
    /// Returns the snapshot taken at reader construction via
    /// `coordinator.all_active_tunnels` filtered to `is_dreamed() == true`.
    /// Matches the Swift `EstateDreamingReader.dreamedActiveTunnels()` which
    /// calls `kit.allActiveTunnels(in:handle)` then filters in-process.
    fn dreamed_active_tunnels(&self) -> Vec<DreamingTunnelItem> {
        self.dreamed_active.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::sync::Arc;

    use genius_locus_kit::coordinator::EstateCoordinator;
    use locus_kit::drawer_store::DrawerStore as LocusDrawerStore;
    use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
    use locus_kit::estate_types::OwnerCredentials;

    fn make_coordinator_and_handle() -> (EstateCoordinator, EstateHandle) {
        let store: Arc<dyn LocusDrawerStore> =
            Arc::new(InMemoryDrawerStore::new(0, None).expect("store"));
        let mut coord = EstateCoordinator::new();
        let handle = coord
            .open(store, OwnerCredentials::new("owner"), 0, 100)
            .expect("open");
        (coord, handle)
    }

    // T8-RDR-1: fresh estate has no dreaming-queue jobs →
    // drain_dreaming_window returns empty.
    #[test]
    fn t8_rdr1_drain_returns_empty_on_fresh_estate() {
        let (coord, handle) = make_coordinator_and_handle();
        let reader = EstateDreamingReader::new(&coord, &handle, "2000-01-01T00:00:00.000Z", "2099-01-01T00:00:00.000Z", 1_000_000.0)
            .expect("new");
        let windows = reader.drain_dreaming_window();
        assert!(windows.is_empty(), "fresh estate has no dreaming-queue jobs");
    }

    // T8-RDR-2: second drain returns empty (drain-once semantics — consumed
    // jobs do not reappear).
    #[test]
    fn t8_rdr2_second_drain_returns_empty() {
        let (coord, handle) = make_coordinator_and_handle();
        let reader = EstateDreamingReader::new(&coord, &handle, "2000-01-01T00:00:00.000Z", "2099-01-01T00:00:00.000Z", 1_000_000.0)
            .expect("new");
        let first = reader.drain_dreaming_window();
        assert!(first.is_empty(), "first drain: empty on fresh estate");
        let second = reader.drain_dreaming_window();
        assert!(second.is_empty(), "second drain: empty (drain-once semantics)");
    }

    // T8-RDR-3: traces snapshot is correct — empty for a fresh estate.
    #[test]
    fn t8_rdr3_traces_empty_on_fresh_estate() {
        let (coord, handle) = make_coordinator_and_handle();
        let reader = EstateDreamingReader::new(&coord, &handle, "2000-01-01T00:00:00.000Z", "2099-01-01T00:00:00.000Z", 1_000_000.0)
            .expect("new");
        let traces = reader.recent_recall_traces();
        assert!(traces.is_empty());
    }

    // T8-RDR-4: tunnels snapshot is correct — empty for a fresh estate.
    #[test]
    fn t8_rdr4_tunnels_empty_on_fresh_estate() {
        let (coord, handle) = make_coordinator_and_handle();
        let reader = EstateDreamingReader::new(&coord, &handle, "2000-01-01T00:00:00.000Z", "2099-01-01T00:00:00.000Z", 1_000_000.0)
            .expect("new");
        let tunnels = reader.existing_tunnels();
        assert!(tunnels.is_empty());
    }
}
